kafka/main.go

```diff
@@ -67,92 +88,145 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
-			consumer := newConsumer(topics, adsConsumerGroupV1, logger)
-			var (
-				failureCount = 0
-				failureLimit = 10
-			)
-			logger.Trace().Msg("Beginning message processing")
-			for {
-				// `FetchMessage` blocks until the next event. Do not block main.
-				ctx := context.Background()
-				logger.Trace().Msgf("Fetching messages from Kafka")
-				msg, err := consumer.FetchMessage(ctx)
-				if err != nil {
-					logger.Error().Err(err).Msg("")
-					if failureCount > failureLimit {
-						break
-					}
-					failureCount++
-					continue
-				}
-				logger.Info().Msgf("Processing message for topic %s at offset %d", msg.Topic, msg.Offset)
-				logger.Info().Msgf("Reader Stats: %#v", consumer.Stats())
-				for _, topicMapping := range topicMappings {
-					if msg.Topic == topicMapping.Topic {
-						go func(
-							msg kafka.Message,
-							topicMapping TopicMapping,
-							providedServer *server.Server,
-							logger *zerolog.Logger,
-						) {
-							err := topicMapping.Processor(
-								msg.Value,
-								topicMapping.ResultProducer,
-								providedServer,
-								logger,
-							)
-							if err != nil {
-								logger.Error().Err(err).Msg("Processing failed.")
-							}
-						}(msg, topicMapping, providedServer, logger)
-
-						if err := consumer.CommitMessages(ctx, msg); err != nil {
-							logger.Error().Msgf("Failed to commit: %s", err)
-						}
-					}
-				}
+	for {
+		msg, err := reader.FetchMessage(ctx)
+		if err != nil {
+			// Indicates batch has no more messages. End the loop for
+			// this batch and fetch another.
+			if err == io.EOF {
+				logger.Info().Msg("Batch complete")
+			} else if strings.ToLower(err.Error()) != "context deadline exceeded" {
+				logger.Error().Err(err).Msg("batch item error")
+				// @TODO: Is there a way to close the batchPipeline and
+				// allow the MessageContexts in it to complete before we
+				// panic? Panic here will only stop this goroutine.
+				panic("failed to fetch kafka messages and closed channel")
 			}
-
-			// The below block will close the producer connection when the error threshold is reached.
-			// @TODO: Test to determine if this Close() impacts the other goroutines that were passed
-			// the same topicMappings before re-enabling this block.
-			//for _, topicMapping := range topicMappings {
-			//	logger.Trace().Msg(fmt.Sprintf("Closing producer connection %v", topicMapping))
-			//	if err := topicMapping.ResultProducer.Close(); err != nil {
-			//		logger.Error().Msg(fmt.Sprintf("Failed to close writer: %e", err))
-			//	}
-			//}
-		}(topicMappings)
+			// There are other possible errors, but the underlying consumer
+			// group handler handle retryable failures well. If further
+			// investigation is needed you can review the handler here:
+			// https://github.com/segmentio/kafka-go/blob/main/consumergroup.go#L729
+			continue
+		}
+		msgCtx := &MessageContext{
+			errorResult: make(chan *utils.ProcessingError),
```

Comment:
can this just be an `error` now?

Reply:
Done. This will definitely help with moving this out into a lib.